src/MOI_wrapper.jl

```diff
@@ -1,58 +1,110 @@
+    # We choose _not_ to support ScalarAffineFunction-in-Interval because Gurobi
+    # introduces some slack variables that makes it hard to keep track.
+    # (MOI.ScalarAffineFunction{Float64},   MOI.Interval{Float64}),
+
+    (MOI.ScalarQuadraticFunction{Float64},  MOI.EqualTo{Float64}),
+    (MOI.ScalarQuadraticFunction{Float64},  MOI.LessThan{Float64}),
+    (MOI.ScalarQuadraticFunction{Float64},  MOI.GreaterThan{Float64})
+)
+
+const SCALAR_SETS = Union{
+    MOI.GreaterThan{Float64}, MOI.LessThan{Float64},
+    MOI.EqualTo{Float64}, MOI.Interval{Float64}
+}
+
+@enum(VariableType, CONTINUOUS, BINARY, INTEGER, SEMIINTEGER, SEMICONTINUOUS)
+@enum(BoundType, NONE, LESS_THAN, GREATER_THAN, LESS_AND_GREATER_THAN, INTERVAL, EQUAL_TO)
+@enum(ObjectiveType, SINGLE_VARIABLE, SCALAR_AFFINE, SCALAR_QUADRATIC)
+
+mutable struct VariableInfo
+    column::Int
+    bound::BoundType
+    type::VariableType
+    name::String
+    # Storage for constraint names associated with variables because Gurobi
+    # can only store names for variables and proper constraints.
+    lessthan_name::String
+    greaterthan_name::String
+    equalto_name::String
+    interval_name::String
+    zeroone_name::String
+    integer_name::String
+    semiinteger_name::String
+    semicontinuous_name::String
+    VariableInfo(column::Int) = new(column, NONE, CONTINUOUS, "", "", "", "", "", "", "", "", "")
+end
+
+mutable struct ConstraintInfo
+    row::Int
+    set::MOI.AbstractSet
+    # Storage for constraint names. Where possible, these are also stored in the
+    # Gurobi model.
+    name::String
+    ConstraintInfo(row::Int, set) = new(row, set, "")
+end
+
+mutable struct Optimizer <: MOI.ModelLike
+    inner::Model
     env::Union{Nothing, Env}
+    silent::Bool
     params::Dict{String, Any}
-    # The next two fields are used to cleverly manage calls to `update_model!`.
+
+    # The next field is used to cleverly manage calls to `update_model!`.
     # `needs_update` is used to record whether an update should be called before
-    # accessing a model attribute (such as the value of a RHS term). One of the
-    # main pain-points when using Gurobi through JuMP is the need to access the
-    # number of variables. Unfortunately, in order to use the Gurobi API (i.e.
-    # `num_vars(model.inner)`), we need to call `update_model!`. To avoid
-    # frequent calls just for this case, we cache the number of variables in
-    # `num_variables`. There are calls in `add_variables` and `delete_variables`
-    # to adjust this value. In addition, we also cache the number of linear and
-    # quadratic constraints.
+    # accessing a model attribute (such as the value of a RHS term).
     needs_update::Bool
-    num_variables::Int
-    num_linear_constraints::Int
-    num_quadratic_constraints::Int
+
+    objective_type::ObjectiveType
+    is_feasibility::Bool
+
+    last_variable_index::Int
+    variable_info::Dict{MOI.VariableIndex, VariableInfo}
```

Comment:
It would be more efficient to have a `Vector{Union{Nothing, AbstractBridge}}` if not too many variables are deleted.
See the `Map` in `src/Bridges/Variable/single_variable_optimizer.jl` in https://github.com/JuliaOpt/MathOptInterface.jl/pull/759.
We could maybe add a dictionary mapping from a bijection of `1:n` with holes to `T` in MOI:
```julia
struct VectorWithHoles{K, V} <: AbstractDict{K, V}
    f::Function # Map index -> key
    values::Vector{Union{Nothing, V}}
end
```
The elements of the dict are `[f(i) => values[i] for i in eachindex(values) if values[i] !== nothing]`.
Here you could use `VectorWithHoles{MOI.VariableIndex, VariableInfo}(MOI.VariableIndex, Union{Nothing, VariableInfo}[])`

Reply:
Re-above: this only makes sense if the `VariableIndex` always begin at `1` after `MOI.empty!`. I'll do some benchmarking to see how the current version compares to LQOI.

Reply:
I do not expect this optimization to be slower than LQOI so this is out of the scope of this PR but it might help with
https://github.com/JuliaOpt/JuMP.jl/issues/1905
In direct mode, it might be a nonnegligible impact on performance